### DISCUSSION TITLE
How does the difficulty work?

### DISCUSSION THREAD

**Discussion by hypr1** [meikpiep/gauguin#74] [General]
Hello! Love the app, it's amazing.

I see there is a range of numbers for each difficulty depending on its size, and each puzzle generated has its own difficulty. How is difficulty determined? Doesn't really seem to be much information. It makes total sense, there's no inherent need to know how difficulty works, but I'd like to.

**Comment by meikpiep** [comment]
Thanks for your question, you're welcome.

The current algorithm is the following: For each cage (group of cells), calculate all possible values to fulfill the cage math (example below). The algorithm takes the amount of possible solutions of each cage. Then, the values of all cages get multiplied.

Example cage: 6x with two cells. This may be solved via (1, 6), (2,3), (3,2) or (6,1). The number of possible solutions is 4.

Only combinations get into account which do not break the Sudoku constraints (there may be no number twice in one row or column). Thus, (2, 2) is no valid solution of a 4x with two cells, but a 4x with three cells which form an edge, may be solved via (2, 1, 2) if the 1 covers the edge itself and the twos are adjacent to it.

As this results in quiet huge values, the natural logarithm (ln(x)) is used to calculate the difficulty.

Though, this covers most games to supply a difficulty rating, it is someway limited. I'm experimenting with a kind of human difficulty rating, where the algorithm uses human techniques to solve Gauguin games. Having each technique mapped to a difficulty rating may lead to a better way of difficulty rating of whole games. But this will be a long time task.

**Comment by hypr1** [comment]
Thank you so much for answering!
